runtime/compiler/env/CHTable.cpp

```diff
@@ -991,6 +1011,136 @@ bool CollectCompiledImplementors::visitSubclass(TR_PersistentClassInfo *cl)
+         // too many classes visited.
+         // set count greater than maxCount, to indicate failure and force exit
+         _count = _maxCount + 1;
+         stopTheWalk();
+         return false;
+         }
+
+      // if the implementor hasn't been cached yet,
+      // add the corresponding subclass to the list
+      TR_ResolvedMethod *resolvedMethod;
+      TR_ResolvedMethodType type = isInterface() ?
+         TR_ResolvedMethodType::Interface :
+         TR_ResolvedMethodType::VirtualFromOffset;
+      auto compInfoPT = static_cast<TR::CompilationInfoPerThreadRemote *>(static_cast<TR_J9VMBase *>(fe())->_compInfoPT);
+      TR_ResolvedMethodKey key =
+         compInfoPT->getResolvedMethodKey(
+            type,
+            reinterpret_cast<TR_OpaqueClassBlock *>(callerMethod->constantPoolHdr()),
+            _slotOrIndex,
+            classId);
+      if (!compInfoPT->getCachedResolvedMethod(
+             key,
+             callerMethod,
+             &resolvedMethod))
+         {
+         _subClasses.push_back(classId);
+         }
+      else if (resolvedMethod)
+         {
+         bool added = addImplementor(resolvedMethod);
+         if (added && _count >= _maxCount)
+            stopTheWalk();
+         }
+      else
+         {
+         // unresolved cached method found.
+         // set count greater than maxCount, to indicate failure and force exit
+         _count = _maxCount + 1;
+         stopTheWalk();
+         return false;
+         }
+      }
+
+   return true;
+   }
+
+void
+CollectResolvedImplementors::cacheResolvedMethods()
+   {
+   TR_ASSERT_FATAL(comp()->isOutOfProcessCompilation(), "Must only be called on JITServer server-side");
+   // _count >= _maxCount indicates that either all required implementors
+   // are already cached or that too many subclasses were visited.
+   // Remote call not needed.
+   if (_count >= _maxCount)
+      {
+      return;
+      }
+
+   auto callerMethod = static_cast<TR_ResolvedJ9JITServerMethod *>(_callerMethod);
+   callerMethod->cacheImplementorMethods(
+      _subClasses,
+      _slotOrIndex,
+      isInterface(),
+      2);
+
+   // now, add the cached methods to _implArray
+   TR_ResolvedMethod *resolvedMethod;
+   TR_ResolvedMethodType type = isInterface() ?
+      TR_ResolvedMethodType::Interface :
+      TR_ResolvedMethodType::VirtualFromOffset;
+   auto compInfoPT = static_cast<TR::CompilationInfoPerThreadRemote *>(static_cast<TR_J9VMBase *>(fe())->_compInfoPT);
+   for (auto it = _subClasses.begin(); it != _subClasses.end(); ++it)
+      {
+      TR_ResolvedMethodKey key =
+         compInfoPT->getResolvedMethodKey(
+            type,
+            reinterpret_cast<TR_OpaqueClassBlock *>(callerMethod->constantPoolHdr()),
+            _slotOrIndex,
+            *it);
+      if (compInfoPT->getCachedResolvedMethod(
+             key,
+             callerMethod,
+             &resolvedMethod))
+         {
+         bool added = addImplementor(resolvedMethod);
```

Comment:
If the cached method is unresolved (`resolvedMethod == NULL`) shouldn't we stop the walk and set `_count == _maxCount`?